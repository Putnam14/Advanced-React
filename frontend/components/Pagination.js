import React from 'react';
import { Query } from 'react-apollo';
import styled from 'styled-components';
import Link from 'next/link';
import PropTypes from 'prop-types';
import { perPage } from '../config';
import { ALL_ITEMS_QUERY } from '../queries/queries.graphql';

const PaginationStyles = styled.div`
  text-align: center;
  display: inline-grid;
  grid-template-columns: repeat(4, auto);
  align-items: stretch;
  justify-content: center;
  align-content: center;
  margin: 2rem 0;
  border: 1px solid ${props => props.theme.lightgrey};
  border-radius: 10px;
  & > * {
    margin: 0;
    padding: 15px 30px;
    border-right: 1px solid ${props => props.theme.lightgrey};
    &:last-child {
      border-right: 0;
    }
  }
  a[aria-disabled='true'] {
    color: grey;
    pointer-events: none;
  }
`;

const Pagination = props => (
  <Query query={ALL_ITEMS_QUERY}>
    {({ data, loading, error }) => {
      if (loading || error) return null;
      const { aggregate } = data.itemsConnection;
      const { page } = props;
      const pages = Math.ceil(aggregate.count / perPage);
      // TODO prefetch this
      return (
        <PaginationStyles data-test="pagination">
          <Link
            prefetch
            href={{
              pathname: 'items',
              query: { page: page - 1 },
            }}
          >
            <a className="prev" aria-disabled={page <= 1}>
              ←Prev
            </a>
          </Link>
          <p>
            Page <strong>{page} </strong> of <strong className="totalPages">{pages}</strong>
          </p>
          <p>
            <strong>{aggregate.count}</strong> Items Total
          </p>
          <Link
            prefetch
            href={{
              pathname: 'items',
              query: { page: page + 1 },
            }}
          >
            <a className="next" aria-disabled={page >= pages}>
              Next →
            </a>
          </Link>
        </PaginationStyles>
      );
    }}
  </Query>
);

Pagination.propTypes = {
  page: PropTypes.number.isRequired,
};

export default Pagination;
