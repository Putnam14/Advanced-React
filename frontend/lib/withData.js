import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from '../queries/queries.graphql';
// can also be a function that accepts a `headers` object (SSR only) and returns a config

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? 'http://localhost:4444' : 'http://localhost:4444',
    ssrMode: !process.browser, // Disables forceFetch on the server (so queries are only run once)
    request: async operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            const { cartOpen } = cache.read({ query: LOCAL_STATE_QUERY });
            const data = { data: { cartOpen: !cartOpen } };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo({ client: createClient });
