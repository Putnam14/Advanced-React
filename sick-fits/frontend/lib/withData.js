// Next-Apollo package to make sure server-side rendering works
import withApollo from 'next-with-apollo'
// You can import Apollo along with any 'link's you want, or do this
import ApolloClient from 'apollo-boost'
import { endpoint } from '../config'

function createClient({ headers }) {
  return new ApolloClient({
    // Can change this if you have different Yoga APIs for dev, prod
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          // Bring cookies along for the ride
          credentials: 'include',
        },
        headers,
      })
    },
  })
}

// Use to create client in _app.js
export default withApollo(createClient)
