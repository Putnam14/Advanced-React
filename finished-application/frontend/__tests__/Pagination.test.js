import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Pagination, { PAGINATION_QUERY } from '../components/Pagination';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';
import { fakeItem } from '../lib/testUtils';

import Router from 'next/router';

// mock the router because there isn't really a router instance
Router.router = { push() {}, prefetch() {} };

function makeMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY, variables: { skip: 0, first: 4 } },
      result: {
        data: {
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

describe('<Pagination/>', () => {
  it('displays loading message', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('div[data-test="pagination"]'))).toMatchSnapshot();
  });

  it('renders pagination for 18 items', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('.totalPages').text()).toEqual('5');
  });

  it('disables prev button on first page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    // first page
    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(true);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
  });

  it('disables next button on last page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={5} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    // first page
    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(true);
  });

  it('enables all buttons on a middle page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={2} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    // first page
    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
  });
});
