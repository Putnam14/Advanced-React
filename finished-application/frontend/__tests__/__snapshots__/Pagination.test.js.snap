// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> displays loading message 1`] = `
<div
  className="sc-bdVaJa ijYKGd"
  data-test="pagination"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 0,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      className="prev"
      href="items?page=0"
      onClick={[Function]}
    >
      ←Prev
    </a>
  </Link>
  <p>
    Page 
    <strong>
      1
       
    </strong>
     of 
    <strong
      className="totalPages"
    >
      1
    </strong>
  </p>
  <p>
    <strong>
      1
    </strong>
     Items Total
  </p>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      className="next"
      href="items?page=2"
      onClick={[Function]}
    >
      Next →
    </a>
  </Link>
</div>
`;
